src/som/interpreter/actors/Actor.java

```diff
@@ -182,11 +187,34 @@ protected void appendToMailbox(final EventualMessage msg) {
     mailboxExtension.append(msg);
   }
 
-  protected static void handleBreakPoints(final EventualMessage msg, final WebDebugger dbg) {
-    if (VmSettings.TRUFFLE_DEBUGGER_ENABLED && msg.isBreakpoint()) {
-      dbg.prepareSteppingUntilNextRootNode();
+  protected static void handleBreakpointsAndStepping(final EventualMessage msg, final WebDebugger dbg) {
+    if (VmSettings.TRUFFLE_DEBUGGER_ENABLED) {
+      // check step-into and breakpoint before the message is processed
+      if (msg.isStepInto()) {
+        dbg.prepareStepIntoMessage();
+      } else if (msg.isBreakpoint()) {
+        dbg.prepareSteppingUntilNextRootNode();
+      }
+      // check if a step-over has been made in the previous message
+      if (stopNextMessage) {
+       dbg.prepareStepIntoMessage();
```

Comment:
after looking at the code, I think all three conditions, this one and above, just want a `dbg.prepareSteppingUntilNextRootNode();`